tests/ui_tools/test_buttons.py

```diff
@@ -23,201 +22,120 @@
-        mocker.patch(STREAMBUTTON + ".mark_muted")
-        show_function = mocker.Mock()
-
-        # To test having more space available with no bullet, but using
-        # same short text, reduce the effective space available
-        if prefix == "":
-            width -= 2
-
-        if isinstance(prefix, tuple):
-            prefix = prefix[1]  # just checking text, not color
-
-        if prefix is None:
-            top_button = TopButton(
-                controller=mocker.Mock(),
-                caption=caption,
-                show_function=show_function,
-                width=width,
-                count=count,
-            )
-            prefix = "\N{BULLET}"
-        else:
-            top_button = TopButton(
-                controller=mocker.Mock(),
-                caption=caption,
-                show_function=show_function,
-                prefix_character=prefix,
-                width=width,
-                count=count,
-            )
-
-        text = top_button._w._original_widget.get_text()
-        count_str = "" if count == 0 else str(count)
-        expected_text = " {}{}{}{}".format(
-            (prefix + " ") if prefix else "",
-            short_text,
-            (width - 2 - (2 if prefix else 0) - len(short_text) - len(count_str)) * " ",
-            count_str,
+    @pytest.fixture(autouse=True)
+    def mock_external_classes(self, mocker):
+        self.controller = mocker.Mock()
+        self.show_function = mocker.Mock()
+        self.urwid = mocker.patch(BUTTONS + ".urwid")
+
+    @pytest.fixture
+    def top_button(self, mocker):
+
+        top_button = TopButton(
+            controller=self.controller,
+            caption="caption",
+            show_function=self.show_function,
+            prefix_character="-",
+            width=0,
+            count=0,
         )
-        assert len(text[0]) == len(expected_text) == (width - 1)
-        assert text[0] == expected_text
 
+        return top_button
+
+    def test_init(self, mocker, top_button):
+
+        assert top_button.controller == self.controller
+        assert top_button._caption == "caption"
+        assert top_button.show_function == self.show_function
+        assert top_button.prefix_character == "-"
+        assert top_button.original_color is None
+        assert top_button.count == 0
+        assert top_button.count_style is None
+
+        # assert top_button.button_prefix == self.urwid.Text("")
+        # assert top_button.button_label == self.urwid.SelectableIcon("")
+        top_button.button_label.set_wrap_mode.assert_called_once_with("ellipsis")
+        assert top_button.button_label.get_cursor_coords("size") is None
+        # assert top_button.button_suffix == self.urwid.Text("")
+
+        self.urwid.Columns.assert_called_once_with(
+            [
+                ("pack", top_button.button_prefix),
+                top_button.button_label,
+                ("pack", top_button.button_suffix),
+            ]
+        )
```

Comment:
I know we have some of these kinds of asserts in other tests, but it'd be good to avoid them if possible, as they basically just mimic the implementation.

I'm wondering if it would be best to replace these kinds of lines with something that compares a render output.

Reply:
But that way we can only test the actual text we are assigning it ( "" for example ) not the actual thing we see (with spacing) as that is handled internally by urwid and depends on width.
Another thing is that this way we can also check the style we are assigning like "selected" for AttrMap.
There is no proper "output" here that we can check against.